File: haskell/day7_part1_2019.hs

import Data.List
import Data.Maybe
import qualified Data.Map as Map
import Control.Concurrent
import Control.Monad

type Memory = Map.Map Int Int
type Program = [Int]
type Address = Int
type Input = Chan Int
type Output = Chan Int

data VM = VM {
    memory :: Memory,
    ip :: Int,
    input :: Input,
    output :: Output
}

opCode :: Int -> Int
opCode cmd = cmd `mod` 100

immediate :: Int -> Int -> Bool
immediate cmd paramNum = (cmd `div` (10 ^ (paramNum + 1))) `mod` 10 == 1

getParam :: VM -> Int -> Bool -> Int
getParam vm addr immediate =
    let param = fromJust $ Map.lookup addr (memory vm)
    in if immediate
        then param
        else fromJust $ Map.lookup param (memory vm)

loadProgram :: String -> IO Program
loadProgram filename = do
    contents <- readFile filename
    return $ map read $ words $ map (\c -> if c == ',' then ' ' else c) contents

runVM :: VM -> IO ()
runVM vm = do
    let cmd = fromJust $ Map.lookup (ip vm) (memory vm)
    case opCode cmd of
        1 -> do
            let param1 = getParam vm (ip vm + 1) (immediate cmd 1)
            let param2 = getParam vm (ip vm + 2) (immediate cmd 2)
            let address = getParam vm (ip vm + 3) True
            let newMemory = Map.insert address (param1 + param2) (memory vm)
            runVM $ vm { memory = newMemory, ip = ip vm + 4 }
        2 -> do
            let param1 = getParam vm (ip vm + 1) (immediate cmd 1)
            let param2 = getParam vm (ip vm + 2) (immediate cmd 2)
            let address = getParam vm (ip vm + 3) True
            let newMemory = Map.insert address (param1 * param2) (memory vm)
            runVM $ vm { memory = newMemory, ip = ip vm + 4 }
        3 -> do
            address <- readChan (input vm)
            let newMemory = Map.insert (getParam vm (ip vm + 1) True) address (memory vm)
            runVM $ vm { memory = newMemory, ip = ip vm + 2 }
        4 -> do
            let param1 = getParam vm (ip vm + 1) (immediate cmd 1)
            writeChan (output vm) param1
            runVM $ vm { ip = ip vm + 2 }
        5 -> do
            let param1 = getParam vm (ip vm + 1) (immediate cmd 1)
            let param2 = getParam vm (ip vm + 2) (immediate cmd 2)
            if param1 /= 0
                then runVM $ vm { ip = param2 }
                else runVM $ vm { ip = ip vm + 3 }
        6 -> do
            let param1 = getParam vm (ip vm + 1) (immediate cmd 1)
            let param2 = getParam vm (ip vm + 2) (immediate cmd 2)
            if param1 == 0
                then runVM $ vm { ip = param2 }
                else runVM $ vm { ip = ip vm + 3 }
        7 -> do
            let param1 = getParam vm (ip vm + 1) (immediate cmd 1)
            let param2 = getParam vm (ip vm + 2) (immediate cmd 2)
            let address = getParam vm (ip vm + 3) True
            let newMemory = Map.insert address (if param1 < param2 then 1 else 0) (memory vm)
            runVM $ vm { memory = newMemory, ip = ip vm + 4 }
        8 -> do
            let param1 = getParam vm (ip vm + 1) (immediate cmd 1)
            let param2 = getParam vm (ip vm + 2) (immediate cmd 2)
            let address = getParam vm (ip vm + 3) True
            let newMemory = Map.insert address (if param1 == param2 then 1 else 0) (memory vm)
            runVM $ vm { memory = newMemory, ip = ip vm + 4 }
        99 -> return ()
        _ -> error $ "Invalid opcode: " ++ show cmd

newVM :: Program -> Input -> Output -> VM
newVM program input output = VM {
    memory = Map.fromList $ zip [0..] program,
    ip = 0,
    input = input,
    output = output
}

runAmplifier :: Program -> [Int] -> IO Int
runAmplifier program phases = do
    chs <- replicateM 6 newChan
    let vms = zipWith3 newVM (repeat program) chs (tail chs)
    forM_ (zip phases vms) $ \(phase, vm) -> do
        writeChan (input vm) phase
        forkIO $ runVM vm
    writeChan (head chs) 0
    readChan (last chs)

main :: IO ()
main = do
    program <- loadProgram "input.txt"
    let phaseSettings = permutations [0, 1, 2, 3, 4]
    results <- mapM (runAmplifier program) phaseSettings
    print $ maximum results

File: r/day7_part1_2019.R

vm <- function(code, input_queue, output_queue) {
  ip <- 1
  get_param <- function(address, immediate) {
    if (immediate) code[address] else code[code[address] + 1]
  }
  is_immediate <- function(cmd, param_num) {
    (cmd %/% 10^(param_num + 1)) %% 10 == 1
  }
  
  while (TRUE) {
    cmd <- code[ip]
    opcode <- cmd %% 100
    
    if (opcode == 1) {
      param1 <- get_param(ip + 1, is_immediate(cmd, 1))
      param2 <- get_param(ip + 2, is_immediate(cmd, 2))
      address <- get_param(ip + 3, TRUE)
      code[address + 1] <- param1 + param2
      ip <- ip + 4
    } else if (opcode == 2) {
      param1 <- get_param(ip + 1, is_immediate(cmd, 1))
      param2 <- get_param(ip + 2, is_immediate(cmd, 2))
      address <- get_param(ip + 3, TRUE)
      code[address + 1] <- param1 * param2
      ip <- ip + 4
    } else if (opcode == 3) {
      address <- get_param(ip + 1, TRUE)
      code[address + 1] <- input_queue[1]
      input_queue <- input_queue[-1]
      ip <- ip + 2
    } else if (opcode == 4) {
      param1 <- get_param(ip + 1, is_immediate(cmd, 1))
      output_queue <- c(output_queue, param1)
      ip <- ip + 2
    } else if (opcode == 5) {
      param1 <- get_param(ip + 1, is_immediate(cmd, 1))
      param2 <- get_param(ip + 2, is_immediate(cmd, 2))
      if (param1 != 0) {
        ip <- param2 + 1
      } else {
        ip <- ip + 3
      }
    } else if (opcode == 6) {
      param1 <- get_param(ip + 1, is_immediate(cmd, 1))
      param2 <- get_param(ip + 2, is_immediate(cmd, 2))
      if (param1 == 0) {
        ip <- param2 + 1
      } else {
        ip <- ip + 3
      }
    } else if (opcode == 7) {
      param1 <- get_param(ip + 1, is_immediate(cmd, 1))
      param2 <- get_param(ip + 2, is_immediate(cmd, 2))
      address <- get_param(ip + 3, TRUE)
      code[address + 1] <- ifelse(param1 < param2, 1, 0)
      ip <- ip + 4
    } else if (opcode == 8) {
      param1 <- get_param(ip + 1, is_immediate(cmd, 1))
      param2 <- get_param(ip + 2, is_immediate(cmd, 2))
      address <- get_param(ip + 3, TRUE)
      code[address + 1] <- ifelse(param1 == param2, 1, 0)
      ip <- ip + 4
    } else if (opcode == 99) {
      return(output_queue)
    }
  }
}

run_amplifiers <- function(phase, code) {
  input <- 0
  for (i in 1:5) {
    input <- vm(code, c(phase[i], input), c())
  }
  return(input)
}

main <- function() {
  code <- scan("input.txt", sep = ",", what = integer())
  
  perms <- function(v) {
    n <- length(v)
    if (n == 1) {
      return(matrix(v, 1, 1))
    }
    res <- matrix(0, nrow = factorial(n), ncol = n)
    idx <- 1
    for (i in 1:n) {
      sub_perms <- perms(v[-i])
      for (j in 1:nrow(sub_perms)) {
        res[idx, ] <- c(v[i], sub_perms[j, ])
        idx <- idx + 1
      }
    }
    return(res)
  }
  
  phase_permutations <- perms(0:4)
  max_output <- 0
  
  for (i in 1:nrow(phase_permutations)) {
    output <- run_amplifiers(phase_permutations[i, ], code)
    max_output <- max(max_output, output)
  }
  
  cat(max_output, "\n")
}

main()
